controller/src/main/java/io/pravega/controller/server/bucket/ZooKeeperBucketManager.java

```diff
@@ -17,33 +17,47 @@
 
 import com.google.common.base.Preconditions;
 import io.pravega.common.Exceptions;
+import io.pravega.controller.store.stream.BucketControllerMap;
 import io.pravega.controller.store.stream.BucketStore;
 import io.pravega.controller.store.stream.ZookeeperBucketStore;
-import io.pravega.controller.util.RetryHelper;
-import lombok.extern.slf4j.Slf4j;
-import org.apache.curator.framework.recipes.cache.PathChildrenCache;
-import org.apache.curator.framework.recipes.cache.PathChildrenCacheListener;
-import org.apache.curator.utils.ZKPaths;
-
+import io.pravega.controller.util.ZKUtils;
 import java.io.IOException;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.Executor;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.function.Function;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.curator.framework.recipes.cache.NodeCache;
+import org.apache.curator.framework.recipes.cache.PathChildrenCache;
+import org.apache.curator.framework.recipes.cache.PathChildrenCacheListener;
+import org.apache.curator.framework.recipes.leader.LeaderSelector;
+import org.apache.curator.utils.ZKPaths;
 
 @SuppressWarnings("deprecation")
 @Slf4j
 public class ZooKeeperBucketManager extends BucketManager {
```

Comment:
Is it possible to make this use composition rather than inheritance? This would involve defining a new interface, but it would simplify reasoning about these two complicated interconnected classes.

Reply:
AbstractService is parent class of BucketManager. We have override AbstractService's method in BucketManager class while zk and inMemory implementation is in respective classes(ZooKeeperBucketManager, InMemoryBucketManager).